libnmstate/netapplier.py

```diff
@@ -63,9 +63,18 @@ def apply(desired_state, verify_change=True, commit=True, rollback_timeout=60):
     validator.validate_dns(desired_state)
     validator.validate_vxlan(desired_state)
 
+    vlans_with_mtu_changes = _prepare_vlan_ifaces_for_mtu_change(desired_state)
     checkpoint = _apply_ifaces_state(
         state.State(desired_state), verify_change, commit, rollback_timeout
     )
+    if vlans_with_mtu_changes:
+        _apply_edited_ifaces_without_checkpoint(
```

Comment:
This code needs to be called inside the `_apply_ifaces_state()` function inside the `with nm.checkpoint.CheckPoint()` context. After the the context is left, the checkpoint might be deleted (commited) but the changes to the MTU should happen while the checkpoint is still present.

Reply:
But `_apply_ifaces_state()` is returning the checkpoint if `commit=False`. If they set `commit=True` it doesn't matter because we are committing the changes after verification phase. Sorry if I am missing something here. Thanks!

Reply:
Let's assume that the user requests a state with an invalid MTU for the VLAN and other changes

If `commit == True` all the other changes will be established on the system and the checkpoint will be removed. In a second step, Nmstate will try to apply the invalid MTU, which will fail. The expected behavior would be to return back to the state of the checkpoint. But this is not possible anymore because the checkpoint is already gone. I wonder, how does the workaround actually work? Doesn't `_prepare_vlan_ifaces_for_mtu_change(desired_state)` need to modify the desired state to remove all VLAN MTU changes from there? Otherwise the verification should fail, shouldn't it?
From my understanding of the issue, the special code needs to be added within the `FIXME` markers in the following listing:

```python
def _apply_ifaces_state(
￼    desired_state, verify_change, commit, rollback_timeout
￼):
￼    current_state = state.State(netinfo.show())
￼
￼    _prepare_desired_state_for_apply(current_state, desired_state)
￼
￼    new_interfaces = _list_new_interfaces(desired_state, current_state)
￼
￼    try:
        with nm.checkpoint.CheckPoint(
            autodestroy=commit, timeout=rollback_timeout
        ) as checkpoint:
            with _setup_providers():
# FIXME: This code needs to be modified to re-apply the vlan state twice
                ifaces2add, ifaces_add_configs = _add_interfaces(
                    new_interfaces, desired_state
                )
                state2edit = _create_editable_desired_state(
                    desired_state, current_state, new_interfaces
                )
                ifaces2edit, ifaces_edit_configs = _edit_interfaces(state2edit)  # this should probably first do all the changes without MTU changes, then also call the next function, afterwards another change with the VLAN MTU changes should be done with _edit_interfaces() and 'set_ifaces_admin_state()
                nm.applier.set_ifaces_admin_state(
                    ifaces2add + ifaces2edit,
                    con_profiles=ifaces_add_configs + ifaces_edit_configs,
                )
# FIXME: END
            if not nmclient.can_disable_ipv6():
                _disable_ipv6(desired_state)
            if verify_change:
                _verify_change(desired_state)  # This should verify the original user-provided state
        if not commit:
            return checkpoint
    except nm.checkpoint.NMCheckPointPermissionError:
        raise NmstatePermissionError('Error creating a check point')
    except nm.checkpoint.NMCheckPointCreationError:
        raise NmstateConflictError('Error creating a check point')
    except NmstateError:
        # Assume rollback occured, revert IPv6 stack state.
        # Checkpoint rollback is async, there is a need to wait for it to
        # finish before proceeding with other actions.
        # TODO: https://nmstate.atlassian.net/browse/NMSTATE-103
        time.sleep(5)
        if not nmclient.can_disable_ipv6():
            _disable_ipv6(current_state)
        raise

```

I might be missing something, so correct me if I am wrong.

Reply:
I agree with @tyll 
@ffmancera do you see a problem with this suggestion?

Reply:
After @tyll clarification I also agree with him. I will work on that. Thanks for explaining!

Reply:
Done! Thanks!